Add a high-level Kafka object that handles request-response

globetro opened the pull request.

dhruvp commented on README.md:
````diff
@@ -16,15 +16,39 @@ Adapter to use Kafka for Udacity Services
+### KafkaAdapter
+#### Create an adapter
 ```javascript
 var KafkaAdapter = require('node-kafka-adapter').KafkaAdapter;
````
should be require('node-kafka-adapter').Kafka now no?

dhruvp commented on src/kafka.js:
```diff
@@ -0,0 +1,96 @@
+  request(serviceName, body, opts={}) {
+    var requestId = this._generateGuid();
+
+    opts = _.extend({
```
from my understanding, requests are supposed to have both a correlation id and a request id

dhruvp commented on src/kafka.js:
```diff
@@ -0,0 +1,96 @@
+    var requestId = this._generateGuid();
+
+    opts = _.extend({
+      correlationId: requestId, // optional
```
I think you want the correlation id and request id to be different. This is because you may fire off multiple requests as a result of an incoming message w/ a given correlation id

globetro replied:
Yep, that's why I provide a way to override the correlation ID if you need to. Otherwise, there's no good reason to make them different.

dhruvp commented on test/kafka_adapter.spec.js:
```diff
@@ -47,21 +47,8 @@ describe('kafka adapter', () => {
     adapter = new KafkaAdapter('fake_zookeeper_address');
   });
   describe('createMessageResponseObject', () => {
-    it('handles malformed requests that are missing correlation_id', function () {
```
why'd you remove this guy? I think correlation id is required at least that's what Art said when I spoke to him last

globetro replied:
Because you have the same test about 20 lines down.

dhruvp replied:
lolz my b

dhruvp commented:
outside of the correlation_id stuff, LGTM. Would just double check w/ Art. Nice work man!

dhruvp commented:
also I wonder if the naming might be consuming for users. As it stands, services would use KafkaAdapter as it has get stream for topic and respondToRequest. Should we maybe move that function over to the Kafka class? Also, how come you removed the dist files? I had that in so that people who weren't using es6 could use this package w.out messing w/ babel. Think that's unnecessary?

globetro commented:
Yeah, I was a bit concerned about naming as well. I was hoping to get it to the point where users would mostly do stuff via the Kafka class, and only use KafkaAdapter to do any low-level stuff. 

I'm considering adding an abstraction for the service-side as well to allow us to do something like: 

```
var kafka = new Kafka('classroom-content');
kafka.listen((request) => {
  return new Promise((resolve, reject) => resolve(graphQl.query(request));
});
```

Thoughts?

dhruvp commented on src/kafka.js:
```diff
@@ -0,0 +1,96 @@
+
+  _initResponseStream() {
+    this.adapter.createTopics(this._getResponseTopic()).then(() => {
+      this.adapter.createMessageStreamForTopic(this._getResponseTopic()).subscribe(
```
oh you probably want to say createMessageStreamForTopic(...).retry().
Rxjs defaults to closing a stream when it catches an error. So unless you say retry you'll never get another response :scream:

dhruvp commented:
I like the idea of users only using the high level Kafka client instead of the adapter. And the method you described above seems nice. One thing to keep in mind is that services will need to create consumers slightly differently. They will need to create the consumer w/ a group id so that they can be associated w/ other instances of that same service listening consuming from that topic.

dhruvp commented on src/kafka.js:
```diff
@@ -0,0 +1,96 @@
+    return `${this.serviceName}.response.${this.id}`;
+  }
+
+  _generateGuid() {
```
Would use more than 2 chars for the guid. At the risk of being douchy here's my reasoning:

If we assume that it's equally probable to land on any 2 letters of the alphabet. There are 676 options for the guid. The probability that there's a collision of guids = 1 - prob(all_different)
if n is the number of instances we spin up
prob(all_different) = (676 * (676-1 ) * ... (676 - n - 1))/(676)^n
At 10 instances that means 
prob (all_different) = 0.935

So the prob of a collision is 0.65 our around 7%

globetro commented:
OK, I've updated to rename Kafka to Client and added a Server. 

As for the services, they should all use the same groupid for the consumer so that they round-robin the responses... so by just not specifying one, it defaults to the same 'kafka-group' group id.

dhruvp commented on src/server.js:
```diff
@@ -0,0 +1,51 @@
+   *
+   * @param  {Function} handler Called with the request body, expects a promise that resolves to a response body
+   */
+  listen(handler) {
```
In the case where an error is thrown it looks like the user of listen won't really get a chance to restart listening.

They'll do something like 

listen(() => {}). At some point if an error is thrown I don't see how they could recover from it.

Might do something like 
```
return Rx.Observable.fromPromise(this._initRequestStream)
  .flatMap((stream) =>  stream)
   .map((request) => Rx.fromPromise(this._processRequest(request. handler)));
```

Then users can do something like
```
responseResultStream = Server.listen(handler).onError(function () {retry logic});
```
Or you could do
```
responseResultStream = Server.listen(handler).retry()...
```

globetro replied:
So I added the retry(). I'm not sure about adding the ability to handle write errors, as I think that should be something hidden from the consumer of this API. You've already implemented retry logic in the write. So at this point, if the write does fail, the Client request will just timeout, which I think is the appropriate overall action.

dhruvp commented on src/server.js:
```diff
@@ -0,0 +1,51 @@
+    return `${this.serviceName}.request`;
+  }
+
+  _initRequestStream() {
```
consider something like this:

```
return Rx.Observable.fromPromise(this.adapter.craeteTopics(...))
   .flatMap(() => this.adapter.createMessageStreamFromTopic(...))
```

This way you can just do 
```
this._initRequestStream().subscribe((request) => console.log(request))
```

globetro replied:
:+1: Thanks for the tip!

dhruvp commented on src/client.js:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  _generateGuid() {
+    return Math.random().toString(36).slice(2);
```
You probably want to bump up the number of chars in the guid to be safe

globetro replied:
If you pass 1 argument to `slice`, it slices starting from there to the end of the string. The reason I'm slicing at 2 is to chop off the '0.' at the beginning (since Math.random() generates a float between 0.0 and 1.0). So this is actually generating a guid like `ocg94hjk4rnewmi`.

I did enjoy all the maths you did to prove why having a 2-digit guid would be a catastrophically moronic idea.

dhruvp replied:
lol. Turns out I'm an idiot haha

dhruvp commented on src/client.js:
```diff
@@ -0,0 +1,96 @@
+
+  _initResponseStream() {
+    this.adapter.createTopics(this._getResponseTopic()).then(() => {
+      this.adapter.createMessageStreamForTopic(this._getResponseTopic()).subscribe(
```
probably want to say createMessageStreamForTopic(...),retry().subscribe...

dhruvp commented on README.md:
```diff
@@ -16,15 +16,63 @@ Adapter to use Kafka for Udacity Services
+var server = new Server('zookeeper:2181', 'classroom-api');
+server.listen(
+  (query) => {
+    return new Promise((resolve, reject) => {
```
can just be graphql.graphql(RootSchema, query)

globetro replied:
Thanks, will update the readme

globetro replied:
So I updated the classroom-content code, but I'm going to leave this example like this as I want to demonstrate that you need to return a promise here.

dhruvp replied:
:+1:

dhruvp commented on src/client.js:
```diff
@@ -0,0 +1,96 @@
+  _initResponseStream() {
+    this.adapter.createTopics(this._getResponseTopic()).then(() => {
+      this.adapter.createMessageStreamForTopic(this._getResponseTopic()).retry().subscribe(
+        (response) => this._processResponse(response)
```
in the case where we get a response that isn't JSON parseable I'm not 100% sure what will happen here. Can we just test for that case and make sure we don't get screwed? I can totally imagine some producer putting something on the stream that isn't proper json

dhruvp commented on src/kafka_adapter.js:
```diff
@@ -48,7 +46,7 @@ export class KafkaAdapter {
   * @param {string} topic The topic name
   * @param {object} optional config to pass in for the consumer
-   *@returns {Observable} 
+   *@returns {Observable}
```
I will use eslint going forward. Sorry lol

dhruvp commented:
ok looks good to me! Thanks for all this! Would look into making sure we handle invalid JSON messages fine also.
